File: drivers/js/src/kv.js
import { RedDBError } from './protocol.js'

export class KvClient {
  constructor(client, collection = 'kv_default') {
    this.client = client
    this.collection = collection
  }

  put(key, value, options = {}) {
    const collection = options.collection ?? this.collection
    const tags = Array.isArray(options.tags) && options.tags.length > 0
      ? ` TAGS [${options.tags.map(kvTagLiteral).join(', ')}]`
      : ''
    const expire = options.expireMs != null ? ` EXPIRE ${Number(options.expireMs)} ms` : ''
    return this.client.call('query', {
      sql: `KV PUT ${kvPath(collection, key)} = ${kvValueLiteral(value)}${expire}${tags}`,
    })
  }

  // Spec-canonical alias for `put` (SDK Helper Spec §5.1 `kv.set`).
  set(key, value, options = {}) {
    return this.put(key, value, options)
  }

  async get(key, options = {}) {
    const collection = options.collection ?? this.collection
    const result = await this.client.call('query', {
      sql: `KV GET ${kvPath(collection, key)}`,
    })
    return result?.rows?.[0]?.value ?? null
  }

  async getMany(keys, options = {}) {
    const values = []
    for (const key of keys) values.push(await this.get(key, options))
    return values
  }

  async exists(key, options = {}) {
    return { exists: (await this.get(key, options)) !== null }
  }

  async delete(key, options = {}) {
    const collection = options.collection ?? this.collection
    const result = await this.client.call('query', {
      sql: `KV DELETE ${kvPath(collection, key)}`,
    })
    const affected = result.affected ?? result.affected_rows ?? 0
    // Spec §5.4 / §2.4 DeleteResult: `deleted` is `affected > 0`.
    return { affected, deleted: affected > 0 }
  }

  async list(options = {}) {
    const collection = options.collection ?? this.collection
    const limit = options.limit == null ? 100 : Number(options.limit)
    if (!Number.isInteger(limit) || limit <= 0) {
      throw new RedDBError('INVALID_ARGUMENT', 'kv.list limit must be a positive integer')
    }
    const prefix = options.prefix == null ? '' : String(options.prefix)
    const result = await this.client.call('query', {
      sql: `SELECT key, value FROM ${kvIdentifier(collection)} ORDER BY key ASC LIMIT ${limit}`,
    })
    const rows = result.rows ?? []
    const items = prefix.length > 0
      ? rows.filter((row) => String(row.key).startsWith(prefix))
      : rows
    return { items }
  }

  async invalidateTags(tags, options = {}) {
    const collection = options.collection ?? this.collection
    const result = await this.client.call('query', {
      sql: `INVALIDATE TAGS [${tags.map(kvTagLiteral).join(', ')}] FROM ${kvIdentifier(collection)}`,
    })
    return result.affected ?? result.affected_rows ?? result.rows?.[0]?.invalidated ?? 0
  }

  async *watch(key, options = {}) {
    if (!this.client.baseUrl) {
      throw new RedDBError('UNSUPPORTED_TRANSPORT', 'kv.watch requires the HTTP transport')
    }
    const collection = options.collection ?? this.collection
    const params = new URLSearchParams()
    if (options.sinceLsn != null) params.set('since_lsn', String(options.sinceLsn))
    if (options.limit != null) params.set('limit', String(options.limit))
    const suffix = params.toString() ? `?${params}` : ''
    const url = `${this.client.baseUrl}/collections/${encodeURIComponent(collection)}/kv/${encodeURIComponent(String(key))}/watch${suffix}`
    const response = await fetch(url, this.client.attachAuth({ method: 'GET' }))
    if (!response.ok) {
      throw new RedDBError('HTTP_ERROR', `kv.watch failed with HTTP ${response.status}`)
    }
    const text = await response.text()
    for (const block of text.split('\n\n')) {
      const line = block.split('\n').find((entry) => entry.startsWith('data: '))
      if (line) yield JSON.parse(line.slice(6))
    }
  }

  watchPrefix(prefix, options = {}) {
    return this.watch(`${prefix}.*`, options)
  }
}

function kvPath(collection, key) {
  return `${kvIdentifier(collection)}.${kvKeySegment(key)}`
}

function kvIdentifier(value) {
  const ident = String(value)
  const invalid = ident.match(/[^A-Za-z0-9_]/)
  if (invalid) {
    throw new RedDBError(
      'INVALID_KV_KEY',
      `invalid KV key "${ident}": character "${invalid[0]}" is not supported`,
    )
  }
  return ident
}

function kvKeySegment(value) {
  const key = String(value)
  if (/^[A-Za-z0-9_]+$/.test(key)) return key
  return `'${key.replace(/'/g, "''")}'`
}

function kvValueLiteral(value) {
  if (typeof value === 'number' || typeof value === 'boolean') return String(value)
  if (value == null) return 'NULL'
  if (typeof value === 'object') return `'${JSON.stringify(value).replace(/'/g, "''")}'`
  return `'${String(value).replace(/'/g, "''")}'`
}

function kvTagLiteral(value) {
  return `'${String(value).replace(/'/g, "''")}'`
}

File: drivers/python-asyncio/src/reddb_asyncio/kv.py
"""KV helpers for the asyncio driver.

Implements the SDK Helper Spec ``kv.*`` surface — exact keys, namespaced
keys (``corpus:version``), object and scalar round-tripping.
"""

from __future__ import annotations

from typing import Any

from .errors import RedDBError
from .sqlutil import sql_identifier


__all__ = [
    "KvClient",
    "kv_path",
    "kv_identifier",
    "kv_key_segment",
    "kv_value_literal",
    "kv_tag_literal",
]


class KvClient:
    """KV namespace bound to a single underlying transport."""

    def __init__(self, transport: Any, collection: str = "kv_default") -> None:
        self._t = transport
        self.collection = collection

    # ------------------------------------------------------------------ spec

    async def set(self, key: str, value: Any, **opts: Any) -> dict[str, Any]:
        return await self.put(key, value, **opts)

    async def put(self, key: str, value: Any, **opts: Any) -> dict[str, Any]:
        collection = opts.pop("collection", self.collection)
        tags = opts.pop("tags", None) or []
        expire_ms = opts.pop("expire_ms", None)
        expire = f" EXPIRE {int(expire_ms)} ms" if expire_ms is not None else ""
        tag_clause = (
            " TAGS [" + ", ".join(kv_tag_literal(tag) for tag in tags) + "]"
            if tags
            else ""
        )
        sql = (
            f"KV PUT {kv_path(collection, key)} = {kv_value_literal(value)}"
            f"{expire}{tag_clause}"
        )
        return await self._t.query(sql)

    async def get(self, key: str, **opts: Any) -> Any:
        collection = opts.pop("collection", self.collection)
        result = await self._t.query(f"KV GET {kv_path(collection, key)}")
        rows = _rows(result)
        if not rows:
            return None
        return rows[0].get("value")

    async def get_many(self, keys: list[str], **opts: Any) -> list[Any]:
        return [await self.get(key, **opts) for key in keys]

    async def exists(self, key: str, **opts: Any) -> dict[str, bool]:
        value = await self.get(key, **opts)
        return {"exists": value is not None}

    async def delete(self, key: str, **opts: Any) -> dict[str, int]:
        collection = opts.pop("collection", self.collection)
        result = await self._t.query(f"KV DELETE {kv_path(collection, key)}")
        affected = 0
        if isinstance(result, dict):
            affected = int(
                result.get("affected")
                or result.get("affected_rows")
                or 0
            )
        return {"affected": affected}

    async def list(self, **opts: Any) -> dict[str, list[dict[str, Any]]]:
        collection = opts.pop("collection", self.collection)
        limit = opts.pop("limit", 100)
        if not isinstance(limit, int) or isinstance(limit, bool) or limit <= 0:
            raise RedDBError(
                "kv.list limit must be a positive integer",
                code="INVALID_ARGUMENT",
            )
        prefix = opts.pop("prefix", None)
        sql = (
            f"SELECT key, value FROM {sql_identifier(collection)} "
            f"ORDER BY key ASC LIMIT {int(limit)}"
        )
        result = await self._t.query(sql)
        rows = _rows(result)
        if prefix is not None and str(prefix) != "":
            prefix_str = str(prefix)
            rows = [r for r in rows if str(r.get("key", "")).startswith(prefix_str)]
        return {"items": rows}

    async def invalidate_tags(self, tags: list[str], **opts: Any) -> int:
        collection = opts.pop("collection", self.collection)
        sql = (
            "INVALIDATE TAGS ["
            + ", ".join(kv_tag_literal(tag) for tag in tags)
            + f"] FROM {sql_identifier(collection)}"
        )
        result = await self._t.query(sql)
        rows = _rows(result)
        if rows:
            return int(rows[0].get("invalidated", 0))
        return int(result.get("affected", 0)) if isinstance(result, dict) else 0

    # ----------------------------------------------------------------- watch

    def watch(self, key: str, **opts: Any):
        collection = opts.pop("collection", self.collection)
        if not hasattr(self._t, "kv_watch"):
            raise RedDBError(
                "kv.watch requires the HTTP transport",
                code="UNSUPPORTED_TRANSPORT",
            )
        return self._t.kv_watch(key, collection=collection, **opts)

    def watch_prefix(self, prefix: str, **opts: Any):
        collection = opts.pop("collection", self.collection)
        if not hasattr(self._t, "kv_watch_prefix"):
            raise RedDBError(
                "kv.watch_prefix requires the HTTP transport",
                code="UNSUPPORTED_TRANSPORT",
            )
        return self._t.kv_watch_prefix(prefix, collection=collection, **opts)


# ---------------------------------------------------------------------------
# Pure helpers (unit-testable without a running server)
# ---------------------------------------------------------------------------


def kv_path(collection: str, key: str) -> str:
    return f"{kv_identifier(collection)}.{kv_key_segment(key)}"


def kv_identifier(value: Any) -> str:
    ident = str(value)
    bad = [ch for ch in ident if not (ch.isalnum() or ch == "_")]
    if bad:
        raise RedDBError(
            f'invalid KV collection "{ident}": character "{bad[0]}" is not supported',
            code="INVALID_KV_KEY",
        )
    return ident


def kv_key_segment(value: Any) -> str:
    key = str(value)
    if key and all(ch.isalnum() or ch == "_" for ch in key):
        return key
    return "'" + key.replace("'", "''") + "'"


def kv_value_literal(value: Any) -> str:
    if value is None:
        return "NULL"
    if isinstance(value, bool):
        return "true" if value else "false"
    if isinstance(value, (int, float)):
        return str(value)
    if isinstance(value, (dict, list, tuple)):
        import json as _json

        return "'" + _json.dumps(value, separators=(",", ":")).replace("'", "''") + "'"
    return "'" + str(value).replace("'", "''") + "'"


def kv_tag_literal(value: Any) -> str:
    return "'" + str(value).replace("'", "''") + "'"


def _rows(result: Any) -> list[dict[str, Any]]:
    if not isinstance(result, dict):
        return []
    rows = result.get("rows")
    return list(rows) if isinstance(rows, list) else []
